### PR TITLE
Fix intraprocess test failure

### PR DESCRIPTION
**Pull request by gerkey**
Make intraprocess more robust with a bounded sleep that checks for the goal condition after publishing (#92).

After discussing the situation with @wjwwood, we agreed that:
* there needs to be some delay between publishing and requiring that the subscriber callback has been invoked, even in the intraprocess request; and
* in lieu of any better API (e.g., waiting on a future associated with the subscriber callback), the best that we can do is to have a bounded sleep loop (which is better than a straight sleep because in the common case we'll exit early, shortening the time required for the test).

So that's what I did here in two places, plus some fixes for use of atomic variables.

### CODE REVIEW

**Comment by gerkey** [comment]
Btw, I tested this fix locally on my machine under the conditions in which I was able to reproduce the failure (https://github.com/ros2/system_tests/issues/92#issuecomment-165561880), and it now works reliably.

File: test_rclcpp/test/test_multithreaded.cpp [hunk 1]
```diff
@@ -50,7 +50,7 @@ static inline void multi_consumer_pub_sub_test(bool intra_process)
     [&counter, &intra_process](test_rclcpp::msg::UInt32::ConstSharedPtr msg,
       const rmw_message_info_t & info) -> void
     {
-      ++counter;
+      counter.fetch_add(1);
```
Comment (jacquelinekay) [inline test_rclcpp/test/test_multithreaded.cpp]: as a side note, I think `fetch_add(1)` and `++counter` are actually equivalent (if they aren't, it's a bug in gcc). Using `fetch_add` makes it more obvious to the reader than the variable is atomic, however. (Same with `store` down below.)

http://en.cppreference.com/w/cpp/atomic/atomic/operator_arith
http://en.cppreference.com/w/cpp/atomic/atomic/operator%3D

Comment (gerkey) [inline test_rclcpp/test/test_multithreaded.cpp]: Ah, very good to know! I'll leave them as is here, but will remember in the future that the shorter syntax is just as good.

Comment (wjwwood) [inline test_rclcpp/test/test_multithreaded.cpp]: Cool, I didn't realize that. TIL

**Comment by wjwwood** [comment]
lgtm

**Pull request merged**